src/main/java/com/google/api/generator/gapic/composer/defaultvalue/DefaultValueComposer.java

```diff
@@ -94,78 +98,100 @@ public static Expr createDefaultValue(
     }
 
     if (methodArg.type().equals(methodArg.field().type())) {
-      return createDefaultValue(methodArg.field());
+      return createValue(methodArg.field(), false, resourceNames, messageTypes, valuePatterns);
     }
 
-    return createDefaultValue(
-        Field.builder().setName(methodArg.name()).setType(methodArg.type()).build());
+    return createValue(Field.builder().setName(methodArg.name()).setType(methodArg.type()).build());
   }
 
-  public static Expr createDefaultValue(Field f) {
-    return createDefaultValue(f, false);
+  public static Expr createValue(Field field) {
+    return createValue(
+        field, false, Collections.emptyMap(), Collections.emptyMap(), Collections.emptyMap());
   }
 
-  static Expr createDefaultValue(Field f, boolean useExplicitInitTypeInGenerics) {
-    if (f.isRepeated()) {
+  public static Expr createValue(
+      Field field,
+      boolean useExplicitInitTypeInGenerics,
+      Map<String, ResourceName> resourceNames,
+      Map<String, Message> messageTypes,
+      Map<String, String> valuePatterns) {
+    if (field.isRepeated()) {
       ConcreteReference.Builder refBuilder =
-          ConcreteReference.builder().setClazz(f.isMap() ? HashMap.class : ArrayList.class);
+          ConcreteReference.builder().setClazz(field.isMap() ? HashMap.class : ArrayList.class);
       if (useExplicitInitTypeInGenerics) {
-        if (f.isMap()) {
-          refBuilder = refBuilder.setGenerics(f.type().reference().generics().subList(0, 2));
+        if (field.isMap()) {
+          refBuilder = refBuilder.setGenerics(field.type().reference().generics().subList(0, 2));
         } else {
-          refBuilder = refBuilder.setGenerics(f.type().reference().generics().get(0));
+          refBuilder = refBuilder.setGenerics(field.type().reference().generics().get(0));
         }
       }
 
       TypeNode newType = TypeNode.withReference(refBuilder.build());
       return NewObjectExpr.builder().setType(newType).setIsGeneric(true).build();
     }
 
-    if (f.isEnum()) {
+    if (field.isEnum()) {
       return MethodInvocationExpr.builder()
-          .setStaticReferenceType(f.type())
+          .setStaticReferenceType(field.type())
           .setMethodName("forNumber")
           .setArguments(
               ValueExpr.withValue(
                   PrimitiveValue.builder().setType(TypeNode.INT).setValue("0").build()))
-          .setReturnType(f.type())
+          .setReturnType(field.type())
           .build();
     }
 
-    if (f.isMessage()) {
+    if (field.isMessage()) {
+      String nestedFieldName = field.name();
+      Map<String, String> nestedValuePatterns = new HashMap<>();
+      for (Map.Entry<String, String> entry : valuePatterns.entrySet()) {
+        if (entry.getKey().startsWith(nestedFieldName + '.')) {
+          nestedValuePatterns.put(
+              entry.getKey().substring(nestedFieldName.length() + 1), entry.getValue());
```

Comment:
We don't consider two or more levels of nesting?

Reply:
we do. createValue and createSimpleMessageBuilderValue call each other recursively